### PR TITLE
Add support for `anchor idl fetch` to work outside anchor workspace

### LINKED ISSUE #1453: cli: anchor idl fetch should work outside workspaces

### PR DESCRIPTION
**Pull request by NBNARADHYA**
# Context
Closes #1453 (Refer to this issue for context)

# Changes
1. `anchor idl fetch` can now be run outside an anchor workspace.
2. If it is run outside workspace, `provider.cluster` option has to be provided.

### CODE REVIEW

**Comment by NBNARADHYA** [comment]
@paul-schaaf I have made changes. Please suggest any changes if they are required.

Also had a question: How is anchor cli tested ? If I need to write tests for my change, how do I go about this?

Thanks

**Comment by paul-schaaf** [comment]
dont think we have to require the `provider.cluster` to be provided. if it's not given and we're outside a workspace, the default cluster from the solana config can be used

re:cli tests we..just test it manually. But we can start adding tests to the CI. Are you comfortable doing that or do you prefer us to do it? imo it's fine to just merge this without testing and then merge a ci change later but if you want to add tests to `tests.yaml` now, feel free!

**Comment by NBNARADHYA** [comment]
@paul-schaaf please suggest any changes!

**Comment by NBNARADHYA** [comment]
@armaniferrante

**Comment by armaniferrante** [comment]
> @armaniferrante

please don't needlessly tag. Someone will review when they have time.

**Comment by paul-schaaf** [comment]
looks like you found the `cli-config` solana crate but why did you copy that code into ours instead of importing the crate?

**Comment by NBNARADHYA** [comment]
> looks like you found the `cli-config` solana crate but why did you copy that code into ours instead of importing the crate?

@paul-schaaf makes more sense! I have made changes

File: cli/src/config.rs [hunk 1]
```diff
@@ -445,6 +447,18 @@ impl FromStr for Config {
...
+        io::Error::new(
+            io::ErrorKind::NotFound,
+            "Default Solana config was not found",
+        )
+    })?;
+    let cli_config = SolanaConfig::load(config_file)?;
+    let url = cli_config.json_rpc_url;
+    Ok(url)
+}
+
```
Comment (paul-schaaf) [inline cli/src/config.rs:461]: more idiomatic way of writing this:
```rust
pub fn get_solana_cfg_url() -> Result<String, io::Error> {
    let config_file = CONFIG_FILE.as_ref().ok_or(io::Error::new(
        io::ErrorKind::NotFound,
        "Default Solana config was not found",
    ))?;
    SolanaConfig::load(config_file).map(|config| config.json_rpc_url)
}
```

File: cli/src/lib.rs [hunk 2]
```diff
@@ -1397,8 +1397,24 @@ pub enum BinVerificationState {
 
 // Fetches an IDL for the given program_id.
 fn fetch_idl(cfg_override: &ConfigOverride, idl_addr: Pubkey) -> Result<Idl> {
-    let cfg = Config::discover(cfg_override)?.expect("Inside a workspace");
-    let url = cluster_url(&cfg);
+    let url = match Config::discover(cfg_override)? {
+        Some(cfg) => cluster_url(&cfg),
+        None => {
+            // If the command is not run inside a workspace,
+            // provider.cluster option has to be provided
```
Comment (paul-schaaf) [inline cli/src/lib.rs:1404]: this comment is no longer true

File: cli/src/lib.rs [hunk 3]
```diff
@@ -1397,8 +1397,24 @@ pub enum BinVerificationState {
...
+            if let Some(cluster) = cfg_override.cluster.clone() {
+                let is_localnet = cluster == Cluster::Localnet;
+                match is_localnet {
+                    // If Cluster is Localnet, default url will be used
+                    true => "http://localhost:8899".to_string(),
+                    false => cluster.url().to_string(),
+                }
```
Comment (paul-schaaf) [inline cli/src/lib.rs:1411]: dont think all this:
```rust
                let is_localnet = cluster == Cluster::Localnet;
                match is_localnet {
                    // If Cluster is Localnet, default url will be used
                    true => "http://localhost:8899".to_string(),
                    false => cluster.url().to_string(),
                }
```
is necessary.

cant you just use 
```rust
cluster.url()
```

**Comment by paul-schaaf** [comment]
pls also add a line to the feature section in the changelog for the unreleased release

**Comment by NBNARADHYA** [comment]
@paul-schaaf hope this is done!

**Comment by NBNARADHYA** [comment]
@paul-schaaf I have already added updated the changelog. I am not able to see “allow edits from maintainers” option on this pull request.

**Comment by paul-schaaf** [comment]
u need to update it again. your change is in the wrong place cause we created a release

**Comment by NBNARADHYA** [comment]
@paul-schaaf done !

**Review by paul-schaaf — APPROVED** [review]

**Comment by paul-schaaf** [comment]
can you sign your commits and force push again?

**Comment by NBNARADHYA** [comment]
@paul-schaaf done!

**Comment by paul-schaaf** [comment]
looks like your commits are still not signed

**Comment by NBNARADHYA** [comment]
@paul-schaaf I can see that they are on github

**Comment by fanatid** [comment]
@NBNARADHYA they are not, if commits are signed then you will see `Verified` label
Screenshot from 2022-03-09 11-08-37

**Comment by NBNARADHYA** [comment]
@fanatid @paul-schaaf I used this command while commiting: `git commit -m "message" -S`. I believe this will sign commits. I am not sure what went wrong. BTW this is the output for `git log --show-signature`.
Screenshot from 2022-03-09 14-15-09

**Comment by NBNARADHYA** [comment]
> 

@fanatid Ok. Now I figured out I have to use GPG keys to sign. But i tried to edit those commits by rebasing but it edits all other commits and signs them all for some reason

**Comment by paul-schaaf** [comment]
we've turned off the signing requirement. Just make sure you have no more conflicts and then Ill merge

**Comment by NBNARADHYA** [comment]
@paul-schaaf Great! Thanks. Just resolved the conflicts

**Comment by paul-schaaf** [comment]
thank you for the contribution!

**Comment by NBNARADHYA** [comment]
@paul-schaaf Thank you!

**Pull request merged**